clients/src/main/java/org/apache/kafka/common/utils/Bytes.java

```diff
@@ -0,0 +1,415 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common.utils;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.util.Comparator;
+
+/**
+ * Utility class that handles byte arrays.
+ *
+ * Its implementation is inspired by Apache Hive, org.apache.hadoop.hbase.util.Bytes
+ */
+public class Bytes {
+
+    /** When we encode strings, we always specify UTF8 encoding */
+    private static final String UTF8_ENCODING = "UTF-8";
+
+    /** When we encode strings, we always specify UTF8 encoding */
+    private static final Charset UTF8_CHARSET = Charset.forName(UTF8_ENCODING);
+
+    /**
+     * A byte array comparator based on lexicograpic ordering.
+     */
+    public final static Comparator<byte[]> BYTES_LEXICO_COMPARATOR = new LexicographicByteArrayComparator();
+
+    public interface ByteArrayComparator extends Comparator<byte[]> {
+
+        int compare(final byte[] buffer1, int offset1, int length1,
+                    final byte[] buffer2, int offset2, int length2);
+    }
+
+    public static class LexicographicByteArrayComparator implements ByteArrayComparator {
```

Comment:
Given the frequency of execution in this area, it be worthwhile to lift some low-level lexical-comparison optimizations from here: https://github.com/google/guava/blob/master/guava%2Fsrc%2Fcom%2Fgoogle%2Fcommon%2Fprimitives%2FUnsignedBytes.java